### PR TITLE
feat: regional holidays using cities

### PR DESCRIPTION
**Pull request by luccamps**
WIP: #346 

After data is normalized we can remove offices_regional_holidays relationship.

### CODE REVIEW

File: app/admin/regional_holiday.rb [hunk 1]
```diff
@@ -15,6 +16,9 @@
     column :offices do |holiday|
       offices_by_holiday(holiday)
     end
+    column :cities do |holiday|
```
Comment (adbatista) [inline app/admin/regional_holiday.rb:19]: `column :cities` does not work?

File: app/admin/regional_holiday.rb [hunk 2]
```diff
@@ -1,10 +1,11 @@
 # frozen_string_literal: true
 
 ActiveAdmin.register RegionalHoliday do
-  permit_params :name, :day, :month, office_ids: []
+  permit_params :name, :day, :month, office_ids: [], city_ids: []
 
   filter :name
   filter :offices, multiple: true, collection: -> { Office.active.order(:city) }
+  filter :cities, multiple: true, collection: -> { City.all.order(:name) }
```
Comment (adbatista) [inline app/admin/regional_holiday.rb:8]: you can do `City.order(:name)`

File: app/admin/regional_holiday.rb [hunk 3]
```diff
@@ -36,6 +43,7 @@
       f.input :day
       f.input :month
       f.input :offices, as: :check_boxes, collection: Office.active.order(:city)
+      f.input :cities, as: :check_boxes, collection: City.all.order(:name)
```
Comment (adbatista) [inline app/admin/regional_holiday.rb:46]: this might list a lot of cities, maybe is better to use a select box

File: spec/factories/cities.rb [hunk 4]
```diff
@@ -2,5 +2,11 @@
   factory :city do
     sequence(:name) {|i| "City Name #{i}" }
     state { create(:state) }
+
+    trait :with_holiday do
+      after :create do |city|
+        city.regional_holidays = create_list(:regional_holiday, 5)
```
Comment (adbatista) [inline spec/factories/cities.rb:8]: is this really creating the association in the db?

File: spec/models/city_spec.rb [hunk 5]
```diff
@@ -3,9 +3,31 @@
...
   context 'Validations' do
     it { is_expected.to validate_presence_of(:name) }
   end
+
+  describe '#holidays' do
+    let!(:city) { create(:city) }
+
+    context 'when there are no holidays' do
+      it 'returns an empty array' do
+        expect(city.holidays).to eq([])
```
Comment (adbatista) [inline spec/models/city_spec.rb:18]:
```suggestion
        expect(city.holidays).to be_empty
```

File: app/admin/regional_holiday.rb (hunk 1 above)
Comment (luccamps) [reply app/admin/regional_holiday.rb]: It does. It's fixed now.

File: app/admin/regional_holiday.rb (hunk 2 above)
Comment (luccamps) [reply app/admin/regional_holiday.rb]: I didn't know it was possible to do it that way. It's fixed now!

File: app/admin/regional_holiday.rb (hunk 3 above)
Comment (luccamps) [reply app/admin/regional_holiday.rb]: Done!

image

File: spec/factories/cities.rb (hunk 4 above)
Comment (luccamps) [reply spec/factories/cities.rb]: It does. If I debug it and execute `city.regional_holidays` it returns all holidays associated with the city. I used the same trait on offices as inspiration.

Comment (adbatista) [reply spec/factories/cities.rb]: ok, just one last check, what happens if you do `city.reload.regional_holidays`

Comment (luccamps) [reply spec/factories/cities.rb]: It shows the associated cities. For a moment I thought I would have to manually save with `city.save` but it wasn't required. I've also tried `City.find(city.id).regional_holidays` and got the holidays as well. It seems like it's indeed persisted.

image

Comment (adbatista) [reply spec/factories/cities.rb]: nice

**Review by adbatista — APPROVED** [review]

**Pull request merged**